src/library/dao/trackdao.cpp

```diff
@@ -1622,14 +1622,16 @@ void TrackDAO::markTrackLocationsAsDeleted(const QString& directory) {
     }
 }
 
-// Look for moved files. Look for files that have been marked as "deleted on disk"
-// and see if another "file" with the same name and filesize exists in the track_locations
-// table. That means the file has moved instead of being deleted outright, and so
-// we can salvage your existing metadata that you have in your DB (like cue points, etc.).
-void TrackDAO::detectMovedFiles(QSet<int>* pTracksMovedSetOld, QSet<int>* pTracksMovedSetNew) {
-    //This function should not start a transaction on it's own!
-    //When it's called from libraryscanner.cpp, there already is a transaction
-    //started!
+// Look for moved files. Look for files that have been marked as
+// "deleted on disk" and see if another "file" with the same name and
+// files size exists in the track_locations table. That means the file has
+// moved instead of being deleted outright, and so we can salvage your
+// existing metadata that you have in your DB (like cue points, etc.).
+bool TrackDAO::detectMovedFiles(QSet<int>* pTracksMovedSetOld,
+        QSet<int>* pTracksMovedSetNew, volatile const bool* pCancel) {
```

Comment:
Unfortunately we have here two encapsulated while loops using query and query2 which makes even a very short time for a single query worse.
The outer while loop takes about 0.2 ms on my recent device with only 15 tracks in library. The inner while loop takes 0.0008 ms for each track in library.  
If we assume to move a the root folder of a not uncommon 10,000 tracks query and query2 will return 10,000 entries. 
Unfortunately we compare each by each and have to do 10,000 x 10,000 compares. 
The resulting rough timing calculations looks like that:  
(0.0008 ms * 10,000 + 0.2 ms) * 10,000 = 82,000 ms 

Unfortunately I cannot backup this calculation by real data by now, but even if I have an error of x 10 it is IMHO worth to keep the cancel since it is measured on a recent device. 

Of cause we should improve this duplicate detection algorithm, but that is IMHO not a 1.12 task since it involves some regression risk. The check for cancel takes 0,0015 ms in the outer loop during my test which adds a 
penalty of 15 ms to a 10,000 track scan -> + 0.02 %